File: app/metrics/ninja.py
"""Invoice Ninja metrics: revenue (net), receivables, VAT, payment behaviour."""

from collections import defaultdict
from datetime import date, timedelta

from app.metrics.dates import add_months, month_start, parse_day, quarter_start

COUNTED = ("sent", "partial", "paid")
OPEN = ("sent", "partial")
MONTHS_BACK = 12


def counted(data: dict) -> list[dict]:
    return [i for i in data.get("invoices", []) if i.get("status") in COUNTED and i.get("date")]


def revenue(data: dict, start: date, end: date) -> float:
    """Net revenue by invoice date: VAT is not revenue."""
    return round(sum(i["net"] for i in counted(data) if start <= parse_day(i["date"]) <= end), 2)


def by_month(data: dict, today: date, months: int = MONTHS_BACK) -> list[dict]:
    """[{"month": "2026-09", "net": 5200.0, "prev": 4100.0}] for the chart widget."""
    result = []
    for back in range(months - 1, -1, -1):
        start = add_months(today, -back)
        end = add_months(start, 1) - timedelta(days=1)
        prev_start = date(start.year - 1, start.month, 1)
        prev_end = add_months(prev_start, 1) - timedelta(days=1)
        result.append({"month": start.isoformat()[:7], "net": revenue(data, start, end),
                       "prev": revenue(data, prev_start, prev_end)})
    return result


def open_invoices(data: dict, today: date) -> list[dict]:
    clients = {c["id"]: c["name"] for c in data.get("clients", [])}
    result = []
    for invoice in data.get("invoices", []):
        if invoice.get("status") not in OPEN or invoice.get("balance", 0) <= 0:
            continue
        due = parse_day(invoice.get("due_date"))
        overdue = (today - due).days if due and due < today else 0
        result.append({**invoice, "client": clients.get(invoice.get("client_id"), "?"), "overdue_days": overdue})
    return sorted(result, key=lambda i: -i["overdue_days"])


def vat_period(today: date, interval: str) -> tuple[date, date]:
    start = quarter_start(today) if interval == "quarterly" else month_start(today)
    months = 3 if interval == "quarterly" else 1
    return start, add_months(start, months) - timedelta(days=1)


def output_vat(data: dict, start: date, end: date, method: str) -> float:
    """VAT owed for a period.

    ist (cash basis):    VAT share of payments received in the period (per client ratio)
    soll (accrual):      VAT of invoices dated in the period
    """
    if method == "soll":
        return round(sum(i["taxes"] for i in counted(data) if start <= parse_day(i["date"]) <= end), 2)

    share: dict = defaultdict(lambda: [0.0, 0.0])
    for invoice in counted(data):
        share[invoice["client_id"]][0] += invoice["taxes"]
        share[invoice["client_id"]][1] += invoice["amount"]

    total = 0.0
    for payment in data.get("payments", []):
        day = parse_day(payment.get("date"))
        if not day or not start <= day <= end:
            continue
        taxes, amount = share.get(payment.get("client_id"), (0.0, 0.0))
        total += payment["amount"] * (taxes / amount if amount else 0)
    return round(total, 2)


def input_vat(data: dict, start: date, end: date) -> float:
    return round(sum(e.get("tax", 0) for e in data.get("expenses", [])
                     if (d := parse_day(e.get("date"))) and start <= d <= end), 2)


def vat_liability(data: dict, today: date, interval: str, method: str) -> dict:
    start, end = vat_period(today, interval)
    out = output_vat(data, start, end, method)
    inp = input_vat(data, start, end)
    return {"start": start.isoformat(), "end": end.isoformat(), "output": out, "input": inp,
            "liability": round(out - inp, 2)}


def shares(data: dict, today: date) -> list[dict]:
    """Revenue share per client over the last 12 months."""
    start = today - timedelta(days=365)
    totals: dict = defaultdict(float)
    for invoice in counted(data):
        if parse_day(invoice["date"]) >= start:
            totals[invoice["client_id"]] += invoice["net"]
    whole = sum(totals.values())
    clients = {c["id"]: c["name"] for c in data.get("clients", [])}
    return sorted(
        [{"client_id": cid, "client": clients.get(cid, "?"), "net": round(v, 2),
          "share": v / whole if whole else 0} for cid, v in totals.items()],
        key=lambda x: -x["share"],
    )


def payment_days(data: dict) -> dict:
    """Average days from invoice to the next payment of the same client (approximation)."""
    by_client: dict = defaultdict(list)
    for payment in data.get("payments", []):
        by_client[payment.get("client_id")].append(parse_day(payment.get("date")))
    result = {}
    for client_id, paid in by_client.items():
        paid = sorted(d for d in paid if d)
        gaps = []
        for invoice in counted(data):
            if invoice["client_id"] != client_id or invoice["status"] != "paid":
                continue
            day = parse_day(invoice["date"])
            later = [d for d in paid if d >= day]
            if later:
                gaps.append((later[0] - day).days)
        if gaps:
            result[client_id] = round(sum(gaps) / len(gaps))
    return result


def summary(data: dict, today: date, interval: str = "monthly", method: str = "ist") -> dict:
    year_start = date(today.year, 1, 1)
    last_year = date(today.year - 1, today.month, today.day) if not (today.month == 2 and today.day == 29) \
        else date(today.year - 1, 2, 28)
    open_items = open_invoices(data, today)
    return {
        "revenue_ytd": revenue(data, year_start, today),
        "revenue_prev_ytd": revenue(data, date(today.year - 1, 1, 1), last_year),
        "revenue_month": revenue(data, month_start(today), today),
        "open_amount": round(sum(i["balance"] for i in open_items), 2),
        "overdue": [i for i in open_items if i["overdue_days"] > 0],
        "open": open_items,
        "vat": vat_liability(data, today, interval, method),
        "currency": data.get("currency", "EUR"),
    }


def forecast_year(data: dict, today: date) -> float:
    """Linear projection of net revenue to Dec 31 from the pace so far."""
    ytd = revenue(data, date(today.year, 1, 1), today)
    elapsed = (today - date(today.year, 1, 1)).days + 1
    days = (date(today.year, 12, 31) - date(today.year, 1, 1)).days + 1
    return round(ytd / elapsed * days, 2)


def cash_expected(data: dict, today: date, days: int) -> float:
    """Receivables plus recurring invoices due within `days` (gross)."""
    horizon = today + timedelta(days=days)
    open_amount = sum(i["balance"] for i in open_invoices(data, today))
    recurring = sum(r["amount"] for r in data.get("recurring", [])
                    if r.get("active") and (d := parse_day(r.get("next_send_date"))) and today <= d <= horizon)
    return round(open_amount + recurring, 2)

File: goapp/internal/metrics/ninja.go
// Package metrics: Invoice Ninja — revenue (net), receivables, VAT, payment behaviour.
package metrics

import (
	"sort"
	"time"

	"dashboard/internal/sources"
)

const (
	monthsBack  = 12
	daysPerYear = 365
)

func isCounted(status string) bool {
	return status == "sent" || status == "partial" || status == "paid"
}
func isOpen(status string) bool { return status == "sent" || status == "partial" }

// NinjaCounted returns invoices counted as revenue (sent/partial/paid, dated).
func NinjaCounted(data *sources.NinjaDataset) []sources.NinjaInvoice {
	var out []sources.NinjaInvoice
	for _, i := range data.Invoices {
		if isCounted(i.Status) && i.Date != "" {
			out = append(out, i)
		}
	}
	return out
}

// NinjaRevenue is net revenue by invoice date in [start, end]: VAT is not revenue.
func NinjaRevenue(data *sources.NinjaDataset, start, end time.Time) float64 {
	var total float64
	for _, i := range NinjaCounted(data) {
		d, ok := ParseDay(i.Date)
		if ok && !d.Before(start) && !d.After(end) {
			total += i.Net
		}
	}
	return round2(total)
}

// NinjaMonth is one month's net revenue plus the same month last year, for
// the chart widget.
type NinjaMonth struct {
	Month string // "2026-09"
	Net   float64
	Prev  float64
}

// NinjaByMonth returns the trailing `months` months (oldest first).
func NinjaByMonth(data *sources.NinjaDataset, today time.Time, months int) []NinjaMonth {
	if months == 0 {
		months = monthsBack
	}
	out := make([]NinjaMonth, 0, months)
	for back := months - 1; back >= 0; back-- {
		start := AddMonths(today, -back)
		end := AddMonths(start, 1).AddDate(0, 0, -1)
		prevStart := time.Date(start.Year()-1, start.Month(), 1, 0, 0, 0, 0, time.UTC)
		prevEnd := AddMonths(prevStart, 1).AddDate(0, 0, -1)
		out = append(out, NinjaMonth{
			Month: start.Format("2006-01"), Net: NinjaRevenue(data, start, end),
			Prev: NinjaRevenue(data, prevStart, prevEnd),
		})
	}
	return out
}

// NinjaOpenInvoice is an open invoice enriched with its client name and
// days overdue.
type NinjaOpenInvoice struct {
	sources.NinjaInvoice
	Client      string
	OverdueDays int
}

// NinjaOpenInvoices returns open (sent/partial, balance > 0) invoices,
// most-overdue first.
func NinjaOpenInvoices(data *sources.NinjaDataset, today time.Time) []NinjaOpenInvoice {
	clients := ninjaClientNames(data)
	var out []NinjaOpenInvoice
	for _, i := range data.Invoices {
		if !isOpen(i.Status) || i.Balance <= 0 {
			continue
		}
		overdue := 0
		if due, ok := ParseDay(i.DueDate); ok && due.Before(today) {
			overdue = int(today.Sub(due).Hours() / 24)
		}
		name := clients[i.ClientID]
		if name == "" {
			name = "?"
		}
		out = append(out, NinjaOpenInvoice{NinjaInvoice: i, Client: name, OverdueDays: overdue})
	}
	sort.Slice(out, func(a, b int) bool { return out[a].OverdueDays > out[b].OverdueDays })
	return out
}

func ninjaClientNames(data *sources.NinjaDataset) map[int64]string {
	out := make(map[int64]string, len(data.Clients))
	for _, c := range data.Clients {
		out[c.ID] = c.Name
	}
	return out
}

// NinjaVATPeriod returns [start, end] of the VAT period containing today.
func NinjaVATPeriod(today time.Time, interval string) (time.Time, time.Time) {
	var start time.Time
	months := 1
	if interval == "quarterly" {
		start, months = QuarterStart(today), 3
	} else {
		start = MonthStart(today)
	}
	return start, AddMonths(start, months).AddDate(0, 0, -1)
}

// NinjaOutputVAT is VAT owed for [start, end].
//
//	soll (accrual):   VAT of invoices dated in the period
//	ist (cash basis): VAT share of payments received in the period (per client ratio)
func NinjaOutputVAT(data *sources.NinjaDataset, start, end time.Time, method string) float64 {
	if method == "soll" {
		var total float64
		for _, i := range NinjaCounted(data) {
			if d, ok := ParseDay(i.Date); ok && !d.Before(start) && !d.After(end) {
				total += i.Taxes
			}
		}
		return round2(total)
	}

	type shareT struct{ taxes, amount float64 }
	share := map[int64]shareT{}
	for _, i := range NinjaCounted(data) {
		s := share[i.ClientID]
		s.taxes += i.Taxes
		s.amount += i.Amount
		share[i.ClientID] = s
	}

	var total float64
	for _, p := range data.Payments {
		d, ok := ParseDay(p.Date)
		if !ok || d.Before(start) || d.After(end) {
			continue
		}
		s := share[p.ClientID]
		if s.amount != 0 {
			total += p.Amount * (s.taxes / s.amount)
		}
	}
	return round2(total)
}

// NinjaInputVAT sums expense VAT in [start, end].
func NinjaInputVAT(data *sources.NinjaDataset, start, end time.Time) float64 {
	var total float64
	for _, e := range data.Expenses {
		if d, ok := ParseDay(e.Date); ok && !d.Before(start) && !d.After(end) {
			total += e.Tax
		}
	}
	return round2(total)
}

// NinjaVATLiability is one period's VAT liability (output − input).
type NinjaVATLiability struct {
	Start, End    string
	Output, Input float64
	Liability     float64
}

// NinjaVATLiabilityOf computes the current period's liability.
func NinjaVATLiabilityOf(data *sources.NinjaDataset, today time.Time, interval, method string) NinjaVATLiability {
	start, end := NinjaVATPeriod(today, interval)
	out, inp := NinjaOutputVAT(data, start, end, method), NinjaInputVAT(data, start, end)
	return NinjaVATLiability{
		Start: start.Format("2006-01-02"), End: end.Format("2006-01-02"),
		Output: out, Input: inp, Liability: round2(out - inp),
	}
}

// NinjaClientShare is one client's revenue share over the trailing year.
type NinjaClientShare struct {
	ClientID int64
	Client   string
	Net      float64
	Share    float64
}

// NinjaShares returns revenue share per client over the trailing 12
// months, largest share first.
func NinjaShares(data *sources.NinjaDataset, today time.Time) []NinjaClientShare {
	start := today.AddDate(0, 0, -365)
	totals := map[int64]float64{}
	for _, i := range NinjaCounted(data) {
		if d, ok := ParseDay(i.Date); ok && !d.Before(start) {
			totals[i.ClientID] += i.Net
		}
	}
	var whole float64
	for _, v := range totals {
		whole += v
	}
	clients := ninjaClientNames(data)
	out := make([]NinjaClientShare, 0, len(totals))
	for cid, v := range totals {
		name := clients[cid]
		if name == "" {
			name = "?"
		}
		share := 0.0
		if whole != 0 {
			share = v / whole
		}
		out = append(out, NinjaClientShare{ClientID: cid, Client: name, Net: round2(v), Share: share})
	}
	sort.Slice(out, func(a, b int) bool { return out[a].Share > out[b].Share })
	return out
}

// NinjaPaymentDays is the average days from invoice to the next payment of
// the same client (approximation), keyed by client id.
func NinjaPaymentDays(data *sources.NinjaDataset) map[int64]int {
	byClient := map[int64][]time.Time{}
	for _, p := range data.Payments {
		if d, ok := ParseDay(p.Date); ok {
			byClient[p.ClientID] = append(byClient[p.ClientID], d)
		}
	}

	result := map[int64]int{}
	for clientID, paid := range byClient {
		sort.Slice(paid, func(a, b int) bool { return paid[a].Before(paid[b]) })
		var gaps []int
		for _, i := range NinjaCounted(data) {
			if i.ClientID != clientID || i.Status != "paid" {
				continue
			}
			d, ok := ParseDay(i.Date)
			if !ok {
				continue
			}
			for _, pd := range paid {
				if !pd.Before(d) {
					gaps = append(gaps, int(pd.Sub(d).Hours()/24))
					break
				}
			}
		}
		if len(gaps) > 0 {
			sum := 0
			for _, g := range gaps {
				sum += g
			}
			result[clientID] = int(round(float64(sum) / float64(len(gaps))))
		}
	}
	return result
}

// NinjaSummary is the Invoice Ninja dashboard's headline numbers.
type NinjaSummary struct {
	RevenueYTD     float64
	RevenuePrevYTD float64
	RevenueMonth   float64
	OpenAmount     float64
	Overdue        []NinjaOpenInvoice
	Open           []NinjaOpenInvoice
	VAT            NinjaVATLiability
	Currency       string
}

// NinjaSummaryOf computes NinjaSummary for today.
func NinjaSummaryOf(data *sources.NinjaDataset, today time.Time, interval, method string) NinjaSummary {
	if interval == "" {
		interval = "monthly"
	}
	if method == "" {
		method = "ist"
	}
	yearStart := time.Date(today.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
	lastYear := time.Date(today.Year()-1, today.Month(), today.Day(), 0, 0, 0, 0, time.UTC)
	if today.Month() == time.February && today.Day() == 29 {
		lastYear = time.Date(today.Year()-1, time.February, 28, 0, 0, 0, 0, time.UTC)
	}

	open := NinjaOpenInvoices(data, today)
	var overdue []NinjaOpenInvoice
	var openAmount float64
	for _, i := range open {
		openAmount += i.Balance
		if i.OverdueDays > 0 {
			overdue = append(overdue, i)
		}
	}

	currency := data.Currency
	if currency == "" {
		currency = "EUR"
	}
	return NinjaSummary{
		RevenueYTD:     NinjaRevenue(data, yearStart, today),
		RevenuePrevYTD: NinjaRevenue(data, time.Date(today.Year()-1, 1, 1, 0, 0, 0, 0, time.UTC), lastYear),
		RevenueMonth:   NinjaRevenue(data, MonthStart(today), today), OpenAmount: round2(openAmount),
		Overdue: overdue, Open: open, VAT: NinjaVATLiabilityOf(data, today, interval, method), Currency: currency,
	}
}

// NinjaForecastYear is a linear projection of net revenue to Dec 31, from
// the pace so far.
func NinjaForecastYear(data *sources.NinjaDataset, today time.Time) float64 {
	yearStart := time.Date(today.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
	ytd := NinjaRevenue(data, yearStart, today)
	elapsed := today.Sub(yearStart).Hours()/24 + 1
	yearEnd := time.Date(today.Year(), 12, 31, 0, 0, 0, 0, time.UTC)
	days := yearEnd.Sub(yearStart).Hours()/24 + 1
	return round2(ytd / elapsed * days)
}

// NinjaCashExpected is receivables plus recurring invoices due within
// `days` (gross).
func NinjaCashExpected(data *sources.NinjaDataset, today time.Time, days int) float64 {
	horizon := today.AddDate(0, 0, days)
	var openAmount float64
	for _, i := range NinjaOpenInvoices(data, today) {
		openAmount += i.Balance
	}
	var recurring float64
	for _, r := range data.Recurring {
		if !r.Active {
			continue
		}
		d, ok := ParseDay(r.NextSendDate)
		if ok && !d.Before(today) && !d.After(horizon) {
			recurring += r.Amount
		}
	}
	return round2(openAmount + recurring)
}
